File: priority.cpp
#include <iostream>
#include <vector>
#include <iomanip>
#include <algorithm>
#include <climits>
using namespace std;
// Function to calculate TAT and WT for Non-Preemptive Priority Scheduling
void nonPreemptivePriority(vector<int> processes, vector<int> burstTime, vector<int> arrivalTime, vector<int> priority) {
    int n = processes.size();
    vector<int> tat(n), wt(n), completionTime(n);
    vector<bool> visited(n, false);
    int currentTime = 0, completed = 0;
    while (completed < n) {
        int highestPriority = -1;
        int minPriority = INT_MAX;
        for (int i = 0; i < n; i++) {
            if (!visited[i] && arrivalTime[i] <= currentTime && priority[i] < minPriority) {
                highestPriority = i;
                minPriority = priority[i];
            }
        }
        if (highestPriority != -1) {
            currentTime += burstTime[highestPriority];
            completionTime[highestPriority] = currentTime;
            tat[highestPriority] = completionTime[highestPriority] - arrivalTime[highestPriority];
            wt[highestPriority] = tat[highestPriority] - burstTime[highestPriority];
            visited[highestPriority] = true;
            completed++;
        } else {
            currentTime++;
        }
    }
    cout << "\nNon-Preemptive Priority Scheduling:\n";
    cout << "Process\tArrival Time\tBurst Time\tPriority\tTurnaround Time\tWaiting Time\n";
    double avgTAT = 0, avgWT = 0;
    for (int i = 0; i < n; i++) {
        cout << "P" << processes[i] << "\t\t" << arrivalTime[i] << "\t\t" << burstTime[i] << "\t\t" << priority[i]
             << "\t\t" << tat[i] << "\t\t" << wt[i] << '\n';
        avgTAT += tat[i];
        avgWT += wt[i];
    }
    cout << "\nAverage Turnaround Time: " << fixed << setprecision(2) << avgTAT / n << endl;
    cout << "Average Waiting Time: " << fixed << setprecision(2) << avgWT / n << endl;
}
// Function to calculate TAT and WT for Preemptive Priority Scheduling
void preemptivePriority(vector<int> processes, vector<int> burstTime, vector<int> arrivalTime, vector<int> priority) {
    int n = processes.size();
    vector<int> remainingTime = burstTime;
    vector<int> tat(n), wt(n), completionTime(n);
    int currentTime = 0, completed = 0;
    int highestPriority = -1;
    while (completed < n) {
        highestPriority = -1;
        int minPriority = INT_MAX;
        for (int i = 0; i < n; i++) {
            if (arrivalTime[i] <= currentTime && remainingTime[i] > 0 && priority[i] < minPriority) {
                highestPriority = i;
                minPriority = priority[i];
            }
        }
        if (highestPriority != -1) {
            currentTime++;
            remainingTime[highestPriority]--;
            if (remainingTime[highestPriority] == 0) {
                completionTime[highestPriority] = currentTime;
                tat[highestPriority] = completionTime[highestPriority] - arrivalTime[highestPriority];
                wt[highestPriority] = tat[highestPriority] - burstTime[highestPriority];
                completed++;
            }
        } else {
            currentTime++;
        }
    }
    cout << "\nPreemptive Priority Scheduling:\n";
    cout << "Process\tArrival Time\tBurst Time\tPriority\tTurnaround Time\tWaiting Time\n";
    double avgTAT = 0, avgWT = 0;
    for (int i = 0; i < n; i++) {
        cout << "P" << processes[i] << "\t\t" << arrivalTime[i] << "\t\t" << burstTime[i] << "\t\t" << priority[i]
             << "\t\t" << tat[i] << "\t\t" << wt[i] << '\n';
        avgTAT += tat[i];
        avgWT += wt[i];
    }

    cout << "\nAverage Turnaround Time: " << fixed << setprecision(2) << avgTAT / n << endl;
    cout << "Average Waiting Time: " << fixed << setprecision(2) << avgWT / n << endl;
}

// Menu-driven program
int main() {
    int n;
    cout << "Enter the number of processes: ";
    cin >> n;

    vector<int> processes(n), burstTime(n), arrivalTime(n), priority(n);
    cout << "Enter Arrival Time, Burst Time, and Priority for each process:\n";
    for (int i = 0; i < n; i++) {
        cout << "Process " << i + 1 << ":\n";
        cout << "Arrival Time: ";
        cin >> arrivalTime[i];
        cout << "Burst Time: ";
        cin >> burstTime[i];
        cout << "Priority (lower number = higher priority): ";
        cin >> priority[i];
        processes[i] = i + 1;
    }

    int choice;
    do {
        cout << "\nMenu:\n";
        cout << "1. Non-Preemptive Priority Scheduling\n";
        cout << "2. Preemptive Priority Scheduling\n";
        cout << "3. Exit\n";
        cout << "Enter your choice: ";
        cin >> choice;

        switch (choice) {
            case 1:
                nonPreemptivePriority(processes, burstTime, arrivalTime, priority);
                break;
            case 2:
                preemptivePriority(processes, burstTime, arrivalTime, priority);
                break;
            case 3:
                cout << "Exiting the program.\n";
                break;
            default:
                cout << "Invalid choice. Please try again.\n";
        }
    } while (choice != 3);

    return 0;
}

File: priority.py
import heapq

def calculate_priority(processes, burst_time, arrival_time, priority, preemptive=True):
    n = len(processes)
    completion_time = [0] * n
    turnaround_time = [0] * n
    waiting_time = [0] * n
    remaining_time = burst_time[:]
    time = 0
    completed = 0
    min_heap = []

    if preemptive:
        # Preemptive Priority Scheduling
        while completed < n:
            # Add all arrived processes to the heap
            for i in range(n):
                if arrival_time[i] <= time and remaining_time[i] > 0:
                    heapq.heappush(min_heap, (priority[i], i))

            if min_heap:
                # Get the process with the highest priority (lowest value)
                _, i = heapq.heappop(min_heap)
                time += 1
                remaining_time[i] -= 1
                if remaining_time[i] == 0:
                    completed += 1
                    completion_time[i] = time
            else:
                time += 1
    else:
        # Non-Preemptive Priority Scheduling
        ready_queue = []
        while completed < n:
            # Add all arrived processes to the ready queue
            for i in range(n):
                if arrival_time[i] <= time and remaining_time[i] > 0:
                    heapq.heappush(ready_queue, (priority[i], i))

            if ready_queue:
                # Get the process with the highest priority (lowest value)
                _, i = heapq.heappop(ready_queue)
                time += burst_time[i]
                completion_time[i] = time
                remaining_time[i] = 0
                completed += 1
            else:
                time += 1

    # Calculate Turnaround Time (TAT) and Waiting Time (WT)
    for i in range(n):
        turnaround_time[i] = completion_time[i] - arrival_time[i]
        waiting_time[i] = turnaround_time[i] - burst_time[i]

    # Display the results
    mode = "Preemptive" if preemptive else "Non-Preemptive"
    print(f"{mode} Priority Scheduling Results:")
    print(f"{'Process':<10}{'Arrival Time':<15}{'Burst Time':<15}{'Priority':<10}{'Completion Time':<20}{'TAT':<10}{'WT':<10}")
    for i in range(n):
        print(f"{processes[i]:<10}{arrival_time[i]:<15}{burst_time[i]:<15}{priority[i]:<10}{completion_time[i]:<20}{turnaround_time[i]:<10}{waiting_time[i]:<10}")
    
    # Calculate and display averages
    avg_tat = sum(turnaround_time) / n
    avg_wt = sum(waiting_time) / n
    print(f"\nAverage Turnaround Time: {avg_tat:.2f}")
    print(f"Average Waiting Time: {avg_wt:.2f}")


# Example Input
processes = ["P1", "P2", "P3", "P4"]
burst_time = [8, 4, 9, 5]
arrival_time = [0, 1, 2, 3]
priority = [2, 1, 4, 3]  # Lower value = Higher priority

# Run Priority Scheduling
print("===" * 10)
calculate_priority(processes, burst_time, arrival_time, priority, preemptive=True)
print("===" * 10)
calculate_priority(processes, burst_time, arrival_time, priority, preemptive=False)
